src/main/java/com/salesforce/storm/spout/dynamic/metrics/TimerManager.java

```diff
@@ -0,0 +1,120 @@
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+ * USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.salesforce.storm.spout.dynamic.metrics;
+
+import org.slf4j.helpers.MessageFormatter;
+
+import java.time.Clock;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Manages start times by key and returns deltas on stop calls for the same key.
+ */
+class TimerManager {
+
+    /**
+     * Keyed start values in millis.
+     */
+    private final Map<String, Long> startValuesMs = new ConcurrentHashMap<>();
+
+    /**
+     * Clock instance, normally a UTC clock but can be overridden for testing.
+     */
+    private Clock clock = Clock.systemUTC();
+
+    /**
+     * Manages start times by key and returns deltas on stop calls for the same key.
+     */
+    TimerManager() {
+    }
+
+    /**
+     * Start a timer based upon a provided key.
+     *
+     * Defaults start time to right now.
+     *
+     * @param key key for timer
+     */
+    void start(final String key) {
+        start(key, getClock().millis());
+    }
+
+    /**
+     * Start a timer based upon a provided key with a specified start time.
+     *
+     * Most likely you don't need this method and should use the one that defaults to right now.
+     *
+     * @param key key for timer
+     * @param startMs time in millis to set as the start
+     */
+    void start(final String key, final long startMs) {
+        if (startValuesMs.containsKey(key)) {
+            throw new IllegalStateException(
+                MessageFormatter.format("The timer key {} already exists in this instances of {}", key, getClass()).getMessage()
+            );
+        }
+
+        startValuesMs.put(key, startMs);
+    }
+
+    /**
+     * Stop a timer based upon a provided key.
+     *
+     * @param key key for timer
+     * @return time elapsed in millis
+     */
+    long stop(final String key) {
+        if (!startValuesMs.containsKey(key)) {
+            throw new IllegalStateException(
+                MessageFormatter.format("The timer key {} does not exist in this instance of {}", key, getClass().toString()).getMessage()
+            );
+        }
+
+        // This gets the value and then removes it, so we could in theory start this key again with a fresh timer
+        final long startTimeMs = startValuesMs.remove(key);
+        final long stopTimeMs = getClock().millis();
+
+        return stopTimeMs - startTimeMs;
+    }
+
+    /**
+     * Get clock instance.
+     * @return clock instance
+     */
+    private Clock getClock() {
+        return this.clock;
+    }
+
+    /**
+     * Override the manager's clock instance, this should only be done in tests.
+     *
+     * @param clock clock instance for testing
+     */
+    void setClock(final Clock clock) {
```

Comment:
Wonder if this should be a package protected constructor?

Reply:
Tried that, but it didn’t work with the test since I need to pass in a new clock on the same instance. Check out the start stop test.